Update README (#55)
ocaml-multicore/lockfree: README.md

| [MPSC Queue](src/mpsc_queue.mli) | A multi-producer, single-consumer, thread-safe queue without support for cancellation. This makes a good data structure for a scheduler's run queue. It is used in [Eio](https://github.com/ocaml-multicore/eio). | It is a single consumer version of the queue described in [Implementing lock-free queues](https://people.cs.pitt.edu/~jacklange/teaching/cs2510-f12/papers/implementing_lock_free.pdf). |

## Lockfree ?
### Definition
[Wikipedia definition](https://en.wikipedia.org/wiki/Non-blocking_algorithm#Lock-freedom) is pretty clear. Briefly (and a bit inaccurately), it says that lock-freedom means:
- obstruction freedom : any domain running in *isolation** for long enough can finish on its own
- guarantees overall progress : at least one domain is doing progress and will finish its call in a finite number of steps.

**In isolation* : a domain is said to run *in isolation* when it is the only domain to take steps. The other domains do nothing during that time even if they are in the middle of a method call. It is used to described properties, but not an actual domain behaviour.

> ### How does it works ?

Reviewer: ```suggestion
### How does it work?
```

The basic idea is that all operations on shared data use `atomics` variables.

### Why use lock free data structures ?
#### In brief
| Pros |  Cons |
|------|-------|
| thread safe | **not** starvation free |
| non blocking and dead-lock safe |  not easily composable (see [reagents](https://github.com/ocaml-multicore/reagents)) |
| usually good performances as it runs in parallel most often than blocking data structures | still subject to contention issues |

> #### What does that means ?

Reviewer: ```suggestion
#### What does that mean?
```

> If you need a *non-blocking* data structure, you are at the right place. However, if you only need a *thread safe* one you may not find what you need here. Especially, lock free data structures does not necessary have better performances than blocking ones. However,  it ensures that is a thread is delayed, that won't delay the other ones (this is the obstruction freedom property)

Reviewer: ```suggestion
If you need a *non-blocking* data structure, you are at the right place. However, if you only need a *thread safe* one you may not find what you need here. Especially, lock free data structures do not necessary have better performances than blocking ones. However, it ensures that if a thread is delayed, that won't delay the other ones (this is the obstruction freedom property).
```

> Also lock freedom is a restraining property, which means it is not easy (and it may not even be possible), to modify or add a function to a lock free data structure while maintaining this property. That explains why most of time the provided data structure are a very simple API.

Reviewer: ```suggestion
Also lock freedom is a restraining property, which means it is not easy (and it may not even be possible), to modify or add a function to a lock free data structure while maintaining this property. That explains why most of time the provided data structures have a very simple API.
```
